subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java

```diff
@@ -37,26 +35,24 @@
     private final SharedJavaInstallationRegistry registry;
     private final JavaToolchainFactory toolchainFactory;
     private final JavaToolchainProvisioningService installService;
-    private final ObjectFactory objectFactory;
 
     @Inject
-    public JavaToolchainQueryService(SharedJavaInstallationRegistry registry, JavaToolchainFactory toolchainFactory, JavaToolchainProvisioningService provisioningService, ObjectFactory objectFactory) {
+    public JavaToolchainQueryService(SharedJavaInstallationRegistry registry, JavaToolchainFactory toolchainFactory, JavaToolchainProvisioningService provisioningService) {
         this.registry = registry;
         this.toolchainFactory = toolchainFactory;
         this.installService = provisioningService;
-        this.objectFactory = objectFactory;
     }
 
-    Provider<JavaCompiler> compilerFrom(Action<? super JavaToolchainSpec> config) {
-        return findMatchingToolchain(configureToolchainSpec(config)).map(JavaToolchain::getJavaCompiler);
+    Provider<JavaCompiler> compilerFor(JavaToolchainSpec spec) {
```

Comment:
I think the internal query service could easily take a Function<Toolchain, Tool> and be tool agnostic. So the default toolchain service uses it as
```
T tool = queryService(spec, JavaToolchain::getJavaLauncher);
```